FOR REVIEW: CPRE DevOps automation changes

toddgiguere opened the pull request:
# Description

This branch represents all changes need to adopt the CPRE DevOps Terraform automation to this module. This Pull Request will allow review and discussion about all proposed changes.

**NOTE:** Applying this PR to main branch will result in a **_BREAKING CHANGE_** and major version bump due to some changes applied. Details in next section.

## Change Highlights
Summary of changes:
- Include all files needed by the DevOps process, including submodule `common-dev-assets`
- Alter Terraform and other files to conform to all format and linting checks performed by pre-commit (*)
- Refactor of unit test using the new ibm-terratest-wrapper library
- Refactor of CI and Release GHA workflows using the new common-pipeline-assets

** (*) BREAKING CHANGE NOTE - ** Most Terraform changes were benign in nature, such as formatting or file naming. However there were some _resource names_ that were changed to pass linting checks. Because the names of resources were changed, applying this new version to existing resources may result in a **destroy/create** operation which could be seen as a breaking change.

toddgiguere committed:
ci: added common-dev-assets submodule

Added common-dev-assets submodule and appropriate symlinks

toddgiguere committed:
refactor: initial pre-commit

Changes after initial run of pre-commit, mostly whitespace fixes

toddgiguere committed:
docs: fixed pre-commit end-of-file with README(s)

toddgiguere committed:
docs: fixed pre-commit terraform_docs in all README(s)

All READMEs were missing the end block comment for the terraform_docs pre-commit hook to properly work

toddgiguere committed:
docs: fixed pre-commit end-of-file after terraform_docs

the terraform_docs pre-commit introduced some more end-of-file issues caught by pre-commit

toddgiguere committed:
refactor: pre-commit fix for terraform_standard_module_structure

Moved all output.tf to outputs.tf, or created empty outputs.tf for examples with no outputs.

To satisfy the terraform_lint of "terraform_standard_module_structure"

toddgiguere committed:
refactor: fix pre-commit for terraform_comment_syntax

Fixed pre-commit warnings for "terraform_comment_syntax". Changed all single line comments from "//" to "#"

toddgiguere committed:
fix: pre-commit fixes for terraform_naming_convention

Fixed all resource names to strictly snake_case to satisfy pre-commit "terraform_naming_convention" rules

BREAKING CHANGE

toddgiguere committed:
fix: pre-commit fixes for terraform_unused_declarations

Fixes any pre-commit errors for terraform_unused_declarations, some by using tf-lint ignore comment, some by removing unused variables

toddgiguere committed:
fix: pre-commit fixed outputs for previous name changes

Needed fixes for a previous commit where module names were slightly changed.
Forgot to change references to new module names in outputs.

toddgiguere committed:
refactor: fixed spelling error of example directory

toddgiguere committed:
fix: pre-commit fixes for terraform_required_providers

Fixes for issues identified by tflint "terraform_required_providers".
Mostly adding version constraint to all required providers, and removing unused provider declarations.

toddgiguere committed:
ci: common-dev-assets merged

Merged in new changes for submodule common-dev-assets

toddgiguere committed:
test: initial setup of tests with testhelper

Initial golang unit test to be run for pull requests and using the ibmcloud-terratest-wrapper/testhelper library

toddgiguere committed:
ci: set up CPRE pipelines and supporting files

Removed old GHA pipelines and replaced with CPRE pipelines, added all necessary additional files used by the CI process, updated common-dev-assets

toddgiguere committed:
refactor: rename of module version file

Renaming the module versions.tf file to version.tf, in order to support future catalog pipeline processing

toddgiguere committed:
docs: added additional readme for tests and workflows

Added some additional README files to explain about test framework and common pipeline usage.

toddgiguere committed:
test: fix regions for test to avoid conflicts

Activity Tracker can only be deployed once per region and account. Changing so that ATR and ATS are deployed to separate regions

ocofaigh commented on .github/workflows/README.md:
````diff
@@ -0,0 +1,22 @@
+  call-terraform-ci-pipeline:
+    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-terraform-module-ci.yml@v1.0.0
+    secrets: inherit
+```
````
This should probably live in the common-pipeline-assets itself?

toddgiguere replied:
This code block represents the usage from the module side of things. For instance, if the module workflow does not include the "secrets: inherit" then the common pipeline will fail.

ocofaigh replied:
I am just thinking of maintaining this. Imagine every repo had to be updated. I see there is now a documentation repo Allen is working on. We could add it there and link it. Just thinking of maintenance going forward.

toddgiguere replied:
Yes thats a good point. I initially had this here so that they would understand it when reviewing this module. But we probably could pull the same file from template and into doc repo.

ocofaigh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,17 @@
+
+jobs:
+  call-terraform-ci-pipeline:
+    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-terraform-module-ci.yml@v1.0.0
```
1.1.0 is out if you want to bump (renovate will bump it anyway once merged)

ocofaigh commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,17 @@
+jobs:
+  call-terraform-release-pipeline:
+    if: ${{ github.event_name == 'workflow_dispatch' || (github.event_name == 'workflow_run' && github.event.workflow_run.conclusion == 'success') }}
+    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-release.yml@v1.0.0
```
1.1.0 is out if you want to bump (renovate will bump it anyway once merged)

ocofaigh commented on modules/activity-tracker-instance/versions.tf:
```diff
@@ -22,7 +22,8 @@ terraform {
   required_providers {
     ibm = {
-      source = "IBM-Cloud/ibm"
+      source  = "IBM-Cloud/ibm"
```
bump the `required_version = ">= 1.0.0"`

toddgiguere replied:
This part is in the comment block, but I can change it so we have a good example.

ocofaigh replied:
`required_version` is not commented out?

toddgiguere replied:
oh good catch, this is in one of the submodules, which I might have missed! I'll double check all of them.

toddgiguere replied:
Fixed the required_version in all of the submodules to match the parent in the latest commit

toddgiguere committed:
chore: change tf required version in submodules to match parent

toddgiguere committed:
docs: submodule doc changes for tf version

ocofaigh commented on .github/workflows/README.md:
````diff
@@ -0,0 +1,22 @@
+  call-terraform-ci-pipeline:
+    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-terraform-module-ci.yml@v1.0.0
+    secrets: inherit
+```
````
Remove this README as it is now covered in https://github.com/terraform-ibm-modules/documentation/blob/main/docs/gh-actions.md and https://github.com/terraform-ibm-modules/common-pipeline-assets/blob/main/README.md

ocofaigh committed:
ci: cleanup

toddgiguere committed:
test: added basic upgrade test

Added the standard UPGRADE test for this module.

SKIP UPGRADE TEST: we do not want this upgrade test to run in this PR, it will run next time

SirSpidey committed:
docs: sync with module template

Update doc files to sync with recent changes to the module template
-  Update tests/README.md to point to project docs
- Update README.md
    - Also make small editorial fixes
- Remove PR template because we have it in the .github repo.

SirSpidey commented:
Comments about the terminology used in the README:
It might make it easier for users if we naming with what's available in the Cloud catalog and docs. 

The diagram aligns pretty well with the catalog (Activity Tracker, Logging (Log Analysis?) and [IBM Cloud] Monitoring). But the readme lists these Activity Tracker and Logging _xTR_ and _xTS_ instances and _a_ Monitoring instance. Are they all supposed to sync with the services listed [here in the docs](https://cloud.ibm.com/docs/activity-tracker?topic=activity-tracker-cloud_services#observability)? If so, why not use those? 

I can't find "Super Tenant" mentioned anywhere in the Cloud Docs. Is there an approved name for the instances? 

Probably don't need the **Acronyms** section. If they remain, ATS, ATR, STR, and STS are defined in the initial section. `OB` isn't used anywhere.

toddgiguere commented:
I realize that the CI pipeline is currently failing after adding an "upgrade" test using our terratest-wrapper. This is due to a bug in the wrapper and is currently being worked on. When fixed I will update the wrapper version and the pipeline tests should pass again.

toddgiguere committed:
test: skip upgrade test

Upgrade test functionality is not working correctly with the new Github Actions pipelines, setting to skip test until resolved

toddgiguere committed:
test: new terratest-wrapper fixes for upgrade test

Trying to enable upgrade tests after fixing bug in the terratest-wrapper go module

SKIP UPGRADE TEST: need to skip upgrade test until PR is merged for first time

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-observability/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
